File: models/action.py
import math
import random
from typing import Optional


class Action:
    DIVIDE = 'DIVIDE'  # деление клетки
    EMIT = 'EMIT'  # выделение вещества
    ABSORB = 'ABSORB'  # поглощение вещества
    MOVE = 'MOVE'  # движение
    HEALS = 'HEALS'  # лечения

    MOVE_RANDOM = 'RANDOM'
    MOVE_TOWARD = 'TOWARD'
    MOVE_AWAY = 'AWAY'
    MOVE_AROUND = 'AROUND'

    def __init__(
            self,
            type_: str,
            power: float = 1.0,
            substance_name: Optional[str] = None,
            move_mode: Optional[str] = None
    ):
        """
        :param type_: тип действия
        :param power: сила действия (энергия, объём, дальность)
        :param substance_name: имя вещества (для EMIT/ABSORB)
        :param move_mode: тип движения: TOWARD / AWAY / RANDOM / AROUND
        """
        self.type = type_
        self.power = power
        self.substance_name = substance_name
        self.move_mode = move_mode

    def execute(self, cell: 'Cell', environment: "Environment"):
        """Выполняет действие"""
        if self.type == Action.DIVIDE:
            self._execute_divide(cell, environment)

        elif self.type == Action.EMIT and self.substance_name:
            self._execute_emit(cell, environment)

        elif self.type == Action.ABSORB and self.substance_name:
            self._execute_absorb(cell, environment)

        elif self.type == Action.MOVE:
            self._execute_move(cell, environment)

        elif self.type == Action.HEALS:
            self._execute_heals(cell)

    def _execute_divide(self, cell: 'Cell', environment: "Environment"):
        new_cell = cell.divide(environment)
        if new_cell:
            environment.add_cell_to_buffer(new_cell)

    def _execute_emit(self, cell: 'Cell', environment: "Environment"):
        cell.emit(self.substance_name, self.power, environment)

    def _execute_absorb(self, cell: 'Cell', environment: "Environment"):
        x, y = cell.get_int_position()
        substance = environment.grid.get_substance(x, y, self.substance_name)
        cell.absorb(substance)

    def _execute_heals(self, cell: 'Cell'):
        cell.heals(self.power)

    def _execute_move(self, cell: 'Cell', environment: "Environment"):
        """Обработка различных режимов движения."""
        if not self.move_mode:
            return

        x, y = cell.get_int_position()

        if self.move_mode == Action.MOVE_RANDOM or not self.substance_name:
            dx = random.uniform(-1, 1)
            dy = random.uniform(-1, 1)

        else:
            # получаем концентрацию в текущей позиции для сравнения
            current_concentration = environment.grid.get_concentration(x, y, self.substance_name)

            # ищем концентрацию вокруг клетки
            best_dir = None
            best_value = None
            vision_radius = 3

            # Проверяем только соседние ячейки (в пределах vision_radius)
            directions = [
                (ix, iy)
                for ix in range(-vision_radius, vision_radius + 1)
                for iy in range(-vision_radius, vision_radius + 1)
                if not (ix == 0 and iy == 0)
            ]

            for (ix, iy) in directions:
                nx, ny = int(x) + ix, int(y) + iy
                # Проверяем границы сетки
                if not (0 <= nx < environment.grid.width and 0 <= ny < environment.grid.height):
                    continue

                val = environment.grid.get_concentration(nx, ny, self.substance_name)

                # Для TOWARD: ищем направление с БОЛЬШЕЙ концентрацией, чем текущая
                if self.move_mode == Action.MOVE_TOWARD:
                    if val > current_concentration:
                        if best_value is None or val > best_value:
                            best_value = val
                            best_dir = (ix, iy)

                # Для AWAY: ищем направление с МЕНЬШЕЙ концентрацией, чем текущая
                elif self.move_mode == Action.MOVE_AWAY:
                    # Если текущая концентрация = 0, ищем направление с минимальной концентрацией
                    if current_concentration == 0:
                        if best_value is None or val < best_value:
                            best_value = val
                            best_dir = (ix, iy)
                    elif val < current_concentration:
                        if best_value is None or val < best_value:
                            best_value = val
                            best_dir = (ix, iy)

                # Для AROUND: сначала находим направление к веществу (как TOWARD)
                elif self.move_mode == Action.MOVE_AROUND:
                    if val > current_concentration:
                        if best_value is None or val > best_value:
                            best_value = val
                            best_dir = (ix, iy)

            # Если не нашли подходящее направление (вещество везде одинаковое или отсутствует)
            if best_dir is None:
                dx = 0
                dy = 0
            else:
                dx, dy = best_dir
                if self.move_mode == Action.MOVE_AROUND:
                    # поворот на 90° против часовой стрелки (перпендикуляр к градиенту)
                    dx, dy = -dy, dx

        # нормализация скорости
        length = math.hypot(dx, dy)
        if length > 0:
            dx = (dx / length) * self.power * 0.1
            dy = (dy / length) * self.power * 0.1
        else:
            # Если длина 0, не двигаемся
            return

        cell.calculate_new_velocity(dx, dy)

    def clone(self) -> 'Action':
        return Action.from_dict(self.to_dict())

    def __repr__(self):
        info = [f"type={self.type}", f"power={self.power:.2f}"]
        if self.substance_name:
            info.append(f"substance={self.substance_name}")
        if self.move_mode:
            info.append(f"move_mode={self.move_mode}")
        return f"Action({', '.join(info)})"

    def to_dict(self):
        return {
            "type": self.type,
            "power": self.power,
            "substance_name": self.substance_name,
            "move_mode": self.move_mode,
        }

    @classmethod
    def from_dict(cls, data):
        return cls(
            type_=data["type"],
            power=data.get("power", 1.0),
            substance_name=data.get("substance_name"),
            move_mode=data.get("move_mode"),
        )

File: java/evolution/src/main/java/com/life/evolution/model/Action.java
package com.life.evolution.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.HashMap;
import java.util.Map;

import static java.lang.Math.hypot;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class Action {

    public static final String DIVIDE = "DIVIDE";
    public static final String EMIT = "EMIT";
    public static final String ABSORB = "ABSORB";
    public static final String MOVE = "MOVE";
    public static final String HEALS = "HEALS";

    public static final String MOVE_RANDOM = "RANDOM";
    public static final String MOVE_TOWARD = "TOWARD";
    public static final String MOVE_AWAY = "AWAY";
    public static final String MOVE_AROUND = "AROUND";

    private String type;
    private double power;
    private String substanceName;
    private String moveMode;

    public Action(String type) {
        this(type, 1.0, null, null);
    }

    // ================================
    // EXECUTE
    // ================================
    public void execute(Cell cell, Environment environment) {
        switch (type) {
            case DIVIDE -> executeDivide(cell, environment);
            case EMIT -> {
                if (substanceName != null) executeEmit(cell, environment);
            }
            case ABSORB -> {
                if (substanceName != null) executeAbsorb(cell, environment);
            }
            case MOVE -> executeMove(cell, environment);
            case HEALS -> executeHeals(cell);
        }
    }

    private void executeDivide(Cell cell, Environment environment) {
        Cell newCell = cell.divide(environment);
        if (newCell != null) environment.addCellToBuffer(newCell);
    }

    private void executeEmit(Cell cell, Environment environment) {
        cell.emit(substanceName, power, environment);
    }

    private void executeAbsorb(Cell cell, Environment environment) {
        int[] pos = cell.getIntPosition();
        Substance s = environment.getGrid().getSubstance(pos[0], pos[1], substanceName);
        cell.absorb(s);
    }

    private void executeHeals(Cell cell) {
        cell.heals(power);
    }

    // ================================
    // MOVEMENT LOGIC
    // ================================
    private void executeMove(Cell cell, Environment environment) {
        if (moveMode == null) return;

        int[] pos = cell.getIntPosition();
        int x = pos[0];
        int y = pos[1];

        double dx = 0;
        double dy = 0;

        if (moveMode.equals(MOVE_RANDOM) || substanceName == null) {
            dx = Math.random() * 2 - 1;
            dy = Math.random() * 2 - 1;

        } else {
            double currentConcentration =
                    environment.getGrid().getConcentration(x, y, substanceName);

            int visionRadius = 3;
            Integer bestIx = null;
            Integer bestIy = null;
            Double bestValue = null;

            for (int ix = -visionRadius; ix <= visionRadius; ix++) {
                for (int iy = -visionRadius; iy <= visionRadius; iy++) {

                    if (ix == 0 && iy == 0) continue;

                    int nx = x + ix;
                    int ny = y + iy;

                    if (nx < 0 || nx >= environment.getGrid().getWidth()) continue;
                    if (ny < 0 || ny >= environment.getGrid().getHeight()) continue;

                    double val = environment.getGrid().getConcentration(nx, ny, substanceName);

                    switch (moveMode) {
                        case MOVE_TOWARD -> {
                            if (val > currentConcentration &&
                                    (bestValue == null || val > bestValue)) {
                                bestValue = val; bestIx = ix; bestIy = iy;
                            }
                        }
                        case MOVE_AWAY -> {
                            if (val < currentConcentration &&
                                    (bestValue == null || val < bestValue)) {
                                bestValue = val; bestIx = ix; bestIy = iy;
                            }
                        }
                        case MOVE_AROUND -> {
                            if (val > currentConcentration &&
                                    (bestValue == null || val > bestValue)) {
                                bestValue = val; bestIx = ix; bestIy = iy;
                            }
                        }
                    }
                }
            }

            if (bestIx == null) {
                dx = dy = 0;
            } else {
                dx = bestIx;
                dy = bestIy;

                if (moveMode.equals(MOVE_AROUND)) {
                    double temp = dx;
                    dx = -dy;
                    dy = temp;
                }
            }
        }

        double length = hypot(dx, dy);
        if (length == 0) return;

        dx = (dx / length) * power * 0.1;
        dy = (dy / length) * power * 0.1;

        cell.calculateNewVelocity(dx, dy);
    }

    // ================================
    // JSON SUPPORT
    // ================================
    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put("type", type);
        map.put("power", power);
        map.put("substance_name", substanceName);
        map.put("move_mode", moveMode);
        return map;
    }

    public static Action fromMap(Map<String, Object> data) {
        return new Action(
                (String) data.get("type"),
                ((Number) data.getOrDefault("power", 1.0)).doubleValue(),
                (String) data.get("substance_name"),
                (String) data.get("move_mode")
        );
    }
}
